toolkit/tools/pkg/imagecustomizerlib/partitionutils.go

```diff
@@ -791,3 +808,153 @@ func getPartitionNum(partitionLoopDevice string) (int, error) {
 
 	return num, nil
 }
+
+// listBtrfsSubvolumes uses btrfs tools to discover actual subvolumes in a mounted BTRFS filesystem
+func listBtrfsSubvolumes(mountPoint string) ([]string, error) {
+	stdout, stderr, err := shell.Execute("btrfs", "subvolume", "list", "-a", mountPoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list BTRFS subvolumes: %w\nstderr: %s", err, stderr)
+	}
+
+	var subvolumes []string
+	lines := strings.Split(strings.TrimSpace(stdout), "\n")
+
+	for _, line := range lines {
+		if strings.TrimSpace(line) == "" {
+			continue
+		}
+
+		// Parse btrfs subvolume list output format:
+		// ID 256 gen 7 top level 5 path @
+		// ID 257 gen 7 top level 5 path @home
+		parts := strings.Fields(line)
+		if len(parts) >= 2 && parts[len(parts)-2] == "path" {
+			subvolPath := parts[len(parts)-1]
+			// Skip special subvolumes that start with <
+			if !strings.HasPrefix(subvolPath, "<") {
+				subvolumes = append(subvolumes, subvolPath)
+			}
+		}
+	}
+
+	logger.Log.Debugf("Found BTRFS subvolumes: %v", subvolumes)
+	return subvolumes, nil
+}
+
+// mountBtrfsSubvolume mounts a specific BTRFS subvolume
+func mountBtrfsSubvolume(devicePath, subvolume, mountPoint string) (*safemount.Mount, error) {
+	mountOptions := fmt.Sprintf("subvol=%s,ro", subvolume)
```

Comment:
In the [mount](https://www.man7.org/linux/man-pages/man2/mount.2.html) syscall, there are integer flags and additional string options. In the [mount](https://www.man7.org/linux/man-pages/man8/mount.8.html) command, all the options are specified as strings, but some are converted into integer flags when sent to the syscall. `ro` is one such flag. It is converted to `unix.MS_RDONLY`, which you already have.